File: antia_numerical_methods/f77/xmp8/plm.f
!     TO EVALUATE ASSOCIATED LEGENDRE FUNCTION

      PROGRAM LEGEND
      IMPLICIT REAL*8(A-H,O-Z)
      DIMENSION P(5000)
 
51    FORMAT('   L =',I6,5X,'M =',I6,5X,'X =',1PD14.6/5X,
     1       'PL(X) =',D14.6,5X,'PLM(X) =',D14.6)

 
100   PRINT *,'TYPE X, L, M      (QUITS WHEN X<-1000)'
      READ *,X,L,M
      IF(X.LT.-1000) STOP

      CALL PLEG(L,X,P)
      F1=P(L+1)
      CALL PLM(L,M,X,P)
      WRITE(6,51) L,M,X,F1,P(L+1)
      GO TO 100
      END
 
!     ----------------------------------------------------
 
!	To calculate Legendre polynomial P_l(X)
!
!	L : (input) Order of polynomial (L.GE.0)
!	X : (input) Argument at which the value of polynomial is required
!	P : (output) Real array of length L+1, which will contain the
!		calculated values of polynomials. P(j+1) will contain P_j(X)
!
!	Required routines : None
 
      SUBROUTINE PLEG(L,X,P)
      IMPLICIT REAL*8(A-H,O-Z)
      DIMENSION P(L+1)
 
      IF(L.LT.0) RETURN
      P(1)=1
      P(2)=X
      DO 2000 N=2,L
        P(N+1)=((2*N-1)*X*P(N)-(N-1)*P(N-1))/N
2000  CONTINUE
      END
 
!     ----------------------------------------------------
 
!	To calculate the associated Legendre functions P_lm(X)
!
!	L,M : (input) Order of function (L.GE.0), ABS(M).LE.L
!	X : (input) Argument at which the value of polynomial is required
!	P : (output) Real array of length L+1, which will contain the
!		calculated values of polynomials. P(j+1) will contain
!		P_jM(X) for j.GE.M
!
!	Required routines : None
 
      SUBROUTINE PLM(L,M,X,P)
      IMPLICIT REAL*8(A-H,O-Z)
      DIMENSION P(L+1)
 
      IF(L.LT.0.OR.ABS(M).GT.L) RETURN
!	First compute P_MM(x)
      MM=ABS(M)
      PM=1.
      DO 2000 I=1,2*MM-1,2
        PM=PM*I
2000  CONTINUE

      IF(M.LT.0) THEN
!	Modify the normalisation factor
        DO 2100 I=1,2*MM
          PM=PM/I
2100    CONTINUE
        IF(MOD(MM,2).EQ.1) PM=-PM
      ENDIF
 
      RM=MM/2.D0
      IF(MM.EQ.0) THEN
        P(MM+1)=1.0
      ELSE IF(ABS(X).LT.1) THEN
        P(MM+1)=PM*(1-X*X)**RM
      ELSE
        P(MM+1)=0.0
      ENDIF

!	Use the recurrence relation to compute P_nM(x)
      P(MM+2)=X*P(MM+1)*(2*MM+1)/(MM+1-M)
      DO 2200 N=MM+2,L
        P(N+1)=((2*N-1)*X*P(N)-(N-1+M)*P(N-1))/(N-M)
2200  CONTINUE
      END

File: antia_numerical_methods/c/xmp/plm.c
/*	To calculate associated Legendre function */

#include <stdio.h>
#include <math.h>

void pleg(int l, double x, double p[]);
void plm(int l, int m, double x, double p[]);

main()
{
	int i,i1,l,m,n;
	double x,theta,phi,f1,f2,p[5000],p1[5000];

	for(i1=0; i1<99; ++i1) {
		printf("type l, m, x,        (quits when x=-9999)\n");
		scanf(" %d %d %le",&l,&m,&x);
		if(x==-9999) return 0;

		pleg(l,x,p);
		plm(l,m,x,p1);
		printf(" l = %d    m = %d   x = %e\n  Pl(x) = %e    Plm(x) = %e\n",l,m,x,p[l],p1[l]);
	}
	return;
}




/*	To calculate Legendre polynomial P_l(X)

	L : (input) Order of polynomial (L>=0)
	X : (input) Argument at which the value of polynomial is required
	P : (output) Array of length L+1, which will contain the
		calculated values of polynomials. P[j] will contain P_j(X)

	Required functions : None
*/

void pleg(int l, double x, double p[])

{
	int n;
	
	if(l<0) return;
	p[0]=1.0;
	p[1]=x;
	for(n=2; n<=l; ++n) p[n]=((2*n-1)*x*p[n-1]-(n-1)*p[n-2])/n;
	return;
}



/*	To calculate the associated Legendre functions P_lm(X)

	L,M : (input) Order of function (L>=0), abs(M)<=L
	X : (input) Argument at which the value of polynomial is required
	P : (output) Array of length L+1, which will contain the
		calculated values of polynomials. P[j] will contain
		P_jM(X) for j>=M

	Required functions : None
*/

#include <math.h>

void plm(int l, int m, double x, double p[])

{
	int i,n,mm;
	double pm,rm;

	if(l<0 || abs(m)>l ) return;
 
/*	First compute P_MM(x) */
	mm=abs(m);
	pm=1.0;
	for(i=1; i<2*mm; i += 2) pm=pm*i;

	if(m<0) {
/*	Modify the normalisation factor */
		for(i=1; i<=2*mm; ++i) pm=pm/i;
		if(mm - 2*(mm/2) == 1) pm=-pm;
	}

	rm=mm/2.0;
	if(mm == 0) p[mm]=1.0;
	else if(fabs(x)<1.0) p[mm]=pm*pow(1.0-x*x, rm);
	else p[mm]=0.0;

/*	Use the recurrence relation to compute P_nM(x) */
	p[mm+1]=x*p[mm]*(2*mm+1)/(mm+1-m);
	for(n=mm+2; n<=l; ++n) p[n]=((2*n-1)*x*p[n-1]-(n-1+m)*p[n-2])/(n-m);

	return;
}
